feat(apigatewayv2): websocket api

ayush987goyal opened the pull request:
feat(apigatewayv2): add support for WebSocket APIs

BREAKING CHANGE: `HttpApiMapping` (and related interfaces for `Attributed` and `Props`) has been renamed to `ApiMapping`
* **apigatewayv2:** `CommonStageOptions` has been renamed to `StageOptions`
* **apigatewayv2:** `HttpStage.fromStageName` has been removed in favour of `HttpStage.fromHttpStageAttributes`
* **apigatewayv2:** `DefaultDomainMappingOptions` has been removed in favour of `DomainMappingOptions`
* **apigatewayv2:** `HttpApiProps.defaultDomainMapping` has been changed from `DefaultDomainMappingOptions` to `DomainMappingOptions`
* **apigatewayv2:** `HttpApi.defaultStage` has been changed from `HttpStage` to `IStage`
* **apigatewayv2:** `IHttpApi.defaultStage` has been removed

closes #2872

Some notes:
1. Only Lambda Integration is currently supported
2. No support for `IntegrationResponse` and `RouteResponse`.
3. The `$default` stageName does not seem to work for WebSocket APIs. Therefore modified the API for defaultStage in the API.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 85c993d7c9cd0395fa778c744b5f259871c61afa
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=533feb1f-c42e-4a97-8b4c-1cfbd4d05d9f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 36f5feb18fd5d52152865de2e9761a73f080a697
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7f8b34cd-5d8a-4199-91d1-751508ec3422%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/websocket/api.ts:
```diff
@@ -0,0 +1,156 @@
+
+  /**
+   * The name of the default stage with deployment
+   * @default - none
```
Does this no default stage will be created? If so, state that.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/websocket/api.ts:
```diff
@@ -0,0 +1,156 @@
+    }
+  }
+
+  /**
+   * @internal
+   */
+  public _addIntegration(config: WebSocketRouteIntegrationConfig): WebSocketIntegration {
```
Wondering if we could move this into the common area for sharing between http and websocket apis. We already have a `common/` folder that is meant to hold some of these.

Maybe we introduce an `ApiBase` private class that is shared between the two? You could also (if needed) introduce an `IApi` interface that is the parent of `IHttpApi` and `IWebsocketApi` that has the `_addIntegration()` method.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/websocket/stage.ts:
```diff
@@ -0,0 +1,67 @@
+ * Represents a stage where an instance of the API is deployed.
+ * @resource AWS::ApiGatewayV2::Stage
+ */
+export class WebSocketStage extends Resource implements IWebSocketStage {
```
Looks so much like our existing http api implementation - https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-apigatewayv2/lib/http/stage.ts. Even the part of API mapping is going to look similar.

I'm mulling over how much of this we should re-use across the two, rather than duplicate. WDYT?

nija-at reviewed (commented):
This is awesome. Thanks for submitting this PR @ayush987goyal.

Some very early feedback. Happy to discuss some of these over chat or on a call, if you prefer.

ayush987goyal commented:
@nija-at I have tried to address most of the comments and taken into account pointers from our discussion. The PR has a lot more number of files now but I believe that will probably be well worth it  in the long run.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 0158d444c84220a301876b7029601e71098c3327
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a9df4f82-7b9a-4ea4-8718-311eae053ecb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 8fca755d07226d7ce14ed74b5ccb2f4f5f2e6fff
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=008fe0d1-8064-44f0-86ef-18b49d77570e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: aceebf0de2c57fbcf8b268088fda806f4e6cf787
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=da21a36e-6821-407a-9c63-47a8dfc46393%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f7bd06616b9729e79ef99de6f3b0628cb49a671d
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=755d8bcc-5308-470c-8b73-9f071da76edc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

ayush987goyal commented:
A couple of extra things to note here are these lint errors which I had to exclude but couldn't find proper way to address. Would love some inputs on them.

```
error: [awslint:construct-interface-extends-iconstruct:@aws-cdk/aws-apigatewayv2.IHttpApi] construct interface must extend core.IConstruct 
error: [awslint:construct-interface-extends-iconstruct:@aws-cdk/aws-apigatewayv2.IWebSocketApi] construct interface must extend core.IConstruct 

error: [awslint:resource-interface-extends-resource:@aws-cdk/aws-apigatewayv2.IHttpApi] construct interfaces of AWS resources must extend cdk.IResource
error: [awslint:resource-interface-extends-resource:@aws-cdk/aws-apigatewayv2.IWebSocketApi] construct interfaces of AWS resources must extend cdk.IResource
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a89e3e5fab66a2cf89d2c30f4ec92a77ba97f34c
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2c243de6-22c4-4ad2-b8b2-c410e90d9111%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:
```diff
@@ -135,7 +72,7 @@ export interface HttpApiProps {
    * @default - no default domain mapping configured. meaningless if `createDefaultStage` is `false`.
    */
-  readonly defaultDomainMapping?: DefaultDomainMappingOptions;
+  readonly defaultDomainMapping?: DomainMappingOptions;
```
I think I understand the change here. 

To clarify I understood this correctly, `DomainMappingOptions` had a `stage` property that was unnecessary because its only usage was within `HttpStageOptions`. Is that correct?

ayush987goyal replied:
Yes that is correct.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:
```diff
@@ -218,45 +155,12 @@ export interface AddRoutesOptions extends BatchHttpRouteOptions {
 }
 
-abstract class HttpApiBase extends Resource implements IHttpApi { // note that this is not exported
+abstract class HttpApiBase extends ApiBase implements IHttpApi { // note that this is not exported
```
Do we need `HttpApiBase` anymore? Can we just use `ApiBase`?

ayush987goyal replied:
We still need it because it has public method `addVpcLink` that can be potentially used for imported APIs as well. To remove it, we will have to duplicate it in the `from*` helpers.

nija-at replied:
Ok.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/integration.ts:
```diff
@@ -9,4 +9,10 @@ export interface IIntegration extends IResource {
\ No newline at end of file
+}
+
+/**
+ * Config representing route integration
+ */
+export interface IRouteIntegrationConfig {
+}
```
Why do we need this? Can we do without?

ayush987goyal replied:
Yes we can. Removed.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/integration.ts:
```diff
@@ -171,13 +171,13 @@ export interface IHttpRouteIntegration {
  * Config returned back as a result of the bind.
  */
-export interface HttpRouteIntegrationConfig {
+export interface IHttpRouteIntegrationConfig extends IRouteIntegrationConfig {
```
We generally don't use `I` prefix for data types but only reserved for construct interfaces. 

In typescript, both complex data types (structs) and interfaces (as defined in the Java world) are called "interfaces". To correctly produce the equivalent structures in other languages and our docs, we use `I` prefix for the latter and not use it for the former.

ayush987goyal replied:
Fixed this post removing `IRouteIntegrationConfig`.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/api.ts:
```diff
@@ -0,0 +1,155 @@
+export abstract class ApiBase extends Resource implements IApi {
+  abstract readonly apiId: string;
+  abstract readonly apiEndpoint: string;
+  protected integrations: Record<string, IIntegration> = {};
```
This should be `@internal` too.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts:
```diff
@@ -1,30 +1,22 @@
 const DEFAULT_STAGE_NAME = '$default';
 
 /**
- * Represents the HttpStage
- */
-export interface IHttpStage extends IStage {
-}
```
Would be nice to keep this in case we want to add methods that are applicable to Http stages but not websocket stages. 

Updating a reference from `IStage` to `IHttpStage` is a breaking change, so we can't add this later.
Worth reviewing all your changes from `IHttpStage` -> `IStage` to see if they should be reverted.

ayush987goyal replied:
So if you look closely, the original implementation of `HttpStage` implemented `IStage` instead of `IHttpStage`. I have added back the `IHttpStage` and made the corresponding changes.

nija-at replied:
> the original implementation of HttpStage implemented IStage instead of IHttpStage

Whoops. That was a miss.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/api.ts:
```diff
@@ -0,0 +1,155 @@
+   */
+  abstract _addIntegration(scope: Construct, config: IRouteIntegrationConfig): IIntegration;
+
+  /**
+   * @internal
+   */
+  protected _getIntegrationConfigHash(scope: Construct, config: IRouteIntegrationConfig): string {
+    const stringifiedConfig = JSON.stringify(Stack.of(scope).resolve(config));
+    const configHash = crypto.createHash('md5').update(stringifiedConfig).digest('hex');
+    return configHash;
+  }
```
Thinking more about this. We could just build a private utility class, say `IntegrationCache`, that holds all of this logic, and remove it entirely from this base class.

The private utility class ca be native typescript so we could do generics, type unions, etc. to get it working best with the two different API classes.

WDYT?

ayush987goyal replied:
Yes that makes sense. Added a utility class under common/private for the same.

nija-at reviewed (dismissed):
I've reviewed the refactors and given the websocket constructs a quick review, but have not yet reviewed the integration classes and module.

These comments should hopefully give you some directions.

This PR will require a few iterations to get right, so please bear with me.

----

> Some experimental interfaces are renamed

You will need to document all of the breaking changes individually, unfortunately. And it will have to be in this particular format - https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#step-4-commit. This is because we use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) to parse commit messages and generate our CHANGELOG.

It's fine if you wait until all the major review iterations are done before fixing this up.

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/api.ts:
```diff
@@ -0,0 +1,155 @@
+    return configHash;
+  }
+
+  /**
+   * @internal
+   */
+  protected _getSavedIntegration(configHash: string) {
+    return this.integrations[configHash];
+  }
+
+  /**
+   * @internal
+   */
+  protected _saveIntegration(configHash: string, integration: IIntegration) {
+    this.integrations[configHash] = integration;
+  }
```
Not applicable now with the IntegrationCache

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/stage.ts:
```diff
@@ -9,22 +14,186 @@ export interface IStage extends IResource {
+/**
+ * Base class representing a Stage
+ */
+export abstract class StageBase extends Resource implements IStage {
```
Yes, done by moving them to base folder.

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/api.ts:
```diff
@@ -0,0 +1,155 @@
+/**
+ * Base class representing an API
+ */
+export abstract class ApiBase extends Resource implements IApi {
```
Yes, done by moving them to base folder.

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/stage.ts:
```diff
@@ -9,22 +14,186 @@ export interface IStage extends IResource {
+  }
+
+  public abstract readonly stageName: string;
+  protected abstract readonly api: IApi;
```
Yeah that should be fine I think. Done.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a066643b7f9f1aa47e49883962676b59f8616027
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7a98a1a3-12b9-433e-8b9a-86dade4b9b4b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
```diff
@@ -146,3 +148,28 @@ const httpEndpoint = new HttpApi(stack, 'HttpProxyPrivateApi', {
+
+## WebSocket APIs
+
+WebSocket integrations connect a route to backend resources. Currently supported integrations include: LambdaWebSocketIntegration
```
Just so we don't have to constantly update this list

```suggestion
WebSocket integrations connect a route to backend resources. The following integrations are supported in the CDK.
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
```diff
@@ -146,3 +148,28 @@ const httpEndpoint = new HttpApi(stack, 'HttpProxyPrivateApi', {
+
+### Lambda WebSocket Integration
+
+Lambda integrations enable integrating a WebSocket API route with a Lambda function. When a client connects/disconnects or sends message specific to a route, the API Gateway service forwards the request to the Lambda function
```
Try to keep line width to 120 or at most 150

```suggestion
Lambda integrations enable integrating a WebSocket API route with a Lambda function. When a client connects/disconnects
or sends message specific to a route, the API Gateway service forwards the request to the Lambda function.
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2-integrations/README.md:
```diff
@@ -146,3 +148,28 @@ const httpEndpoint = new HttpApi(stack, 'HttpProxyPrivateApi', {
+
+Lambda integrations enable integrating a WebSocket API route with a Lambda function. When a client connects/disconnects or sends message specific to a route, the API Gateway service forwards the request to the Lambda function
+
+The API Gateway service will invoke the lambda function with an event payload of a specific format.
```
Any details we can link to on what the lambda event request will contain and what the schema of the response should be.

ayush987goyal replied:
I looked around the API reference and developer docs but was not able to find anything unfortunately. We can add it in later changes once we have one.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/websocket/api.ts:
```diff
@@ -0,0 +1,164 @@
+   */
+  readonly routeSelectionExpression?: string;
+
+  /**
+   * The name of the default stage with deployment
+   * @default - No default stage is created
+   */
+  readonly defaultStageName?: string;
+
```
Why not follow the same pattern as HttpApi here - `defaultStage` and `createDefaultStage` properties? (sidenote: for the latter, I would prefer calling it `disableDefaultStage`)

ayush987goyal replied:
Same as the previous comment regarding default stage name not working in WebSocket APIs

nija-at replied:
I don't think I follow this. What is the difference between websocket and http apis when it comes to default stages? Can you expand?

Comparing the documentation for [websocket api](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-set-up-websocket-deployment.html) and [http api](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-stages.html), perhaps the concept of a default stage doesn't exist on the former? In such a case, should we just drop support for this.

ayush987goyal replied:
So when I tried out the similar flow of `$default` stage for websocket api, there were some problems with the functionality. The cfn deployment was successful but the actual functionality was not working. I think it has to do with the name of the stage (probably containing `$`). I could not find any documentation around the same and hence decided to change the api a bit accordingly.

I agree on dropping the support or maybe diving deep with the apigw team regarding the same. Thoughts?

nija-at replied:
Let's drop support for now. This can always be added later by you or anyone else with a proper deep dive.

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/websocket/api.ts:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    if (!props?.defaultStageName && props?.defaultDomainMapping) {
+      throw new Error('defaultDomainMapping not supported when defaultStageName is not provided',
+      );
```
```suggestion
      throw new Error('defaultDomainMapping not supported when defaultStageName is not provided');
```

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/package.json:
```diff
@@ -103,13 +103,20 @@
   "awslint": {
     "exclude": [
+      "props-physical-name:@aws-cdk/aws-apigatewayv2.ApiMappingProps",
+      "construct-interface-extends-iconstruct:@aws-cdk/aws-apigatewayv2.IHttpApi",
+      "construct-interface-extends-iconstruct:@aws-cdk/aws-apigatewayv2.IWebSocketApi",
+      "resource-interface-extends-resource:@aws-cdk/aws-apigatewayv2.IHttpApi",
+      "resource-interface-extends-resource:@aws-cdk/aws-apigatewayv2.IWebSocketApi",
```
That's odd. Is this because these interfaces don't directly inherit `IConstruct` and `IResource`? 
Can you check if something is amiss in this code? Otherwise, I'll have to look into the linter to see if there's a bug there.

ayush987goyal replied:
Yes I am confused with this as well. Here the reference to `Resource` comes from the base `ApiBase` class and nowhere around `IApi` or `IHttpApi`. Not sure how to address this or maybe a linter issue indeed.

nija-at replied:
I think the problem is that `IApi` does not extend `IResource`. Can you see if that fixes it?

ayush987goyal replied:
Yep that does it. Thanks!

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/api.ts:
```diff
@@ -349,7 +254,7 @@ export class HttpApi extends HttpApiBase {
    * default stage of the api resource
    */
-  public readonly defaultStage: HttpStage | undefined;
+  public readonly defaultStage: IStage | undefined;
```
Similar comment to another one. I think there's value in going from IHttpApi -> IHttpStage.
If not, customers will have to do type checking to correctly get to `IHttpStage`.

Perhaps we should move this down to the subclasses. I'm curious what the implications of that would be.

ayush987goyal replied:
Moving this down to subclasses will not be possible because of how `ApiMapping` works with the new implementation. It requires stage to be present for it's functionalities. So either we create a specific `ApiMapping` per API type (like we had before) or we keep it like so.

nija-at replied:
Could we do something like this instead?

```ts
public get defaultStage(): IHttpStage | undefined {
  if (_defaultStage && !(_defaultStage instanceof IHttpStage)) {
    throw new Error('unexpected error: defaultStage is not of type IHttpStage');
  }
  return _defaultStage;
}
```

The error condition should never occur unless there's a bug within the CDK.

nija-at replied:
Actually, can you explain your previous comment? I can't quite figure out, from the code diffs, why this is necessary for `ApiMapping`.

ayush987goyal replied:
So the `ApiMapping` requires `api` to be passed as props. This `api` with the new implementation takes the form of `IApi` since it is common for both API types. In the implementation of `ApiMapping` we use `defaultStage` if a `stage` is not provided in the props. Now, if we move the `defaultStage` declaration from `IApi` down to individual subclasses, we will not be able to do this. Let me know if that was clear. Also, I am not sure how the above recommendation would pan out. Do we declare this getter in `IApi` too?

nija-at replied:
Given that default stage is a concept that only (seems to) exist on Http APIs, we should bring this down to the HttpApi subclass. Specifically, the logic of falling back to the default stage if a stage is not provided should be specific to `HttpApi`.

ayush987goyal replied:
Changed the implementation in a non breaking way for `ApiMapping` (although the rename from `HttpApiMapping` would still be a breaking change)

nija-at reviewed (dismissed):
This is fantastic!
Just 1 or 2 comments around the design/refactor, the rest are just small improvements and clarifications.

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/stage.ts:
```diff
@@ -9,22 +12,120 @@ export interface IStage extends IResource {
+
+  /**
+   * The API this stage is associated to.
+   * @attribute
```
Got it. Thanks for clarification!

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts:
```diff
@@ -16,15 +15,14 @@ export interface IHttpStage extends IStage {
+   * @default '$default' the default stage of the API. This stage will have the URL at the root of the API endpoint.
    */
-  readonly domainMapping?: DomainMappingOptions;
+  readonly stageName?: string;
```
As mentioned in the PR description:
> The $default stageName does not seem to work for WebSocket APIs. Therefore modified the API for defaultStage in the API.

The `stageName` is a required prop for `WebSocketStage` but not `HttpStage`. Therefore it is not part of common options.

nija-at replied:
I don't think I follow this. What is the difference between websocket and http apis when it comes to default stages? Can you expand?

Comparing the documentation for [websocket api](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-set-up-websocket-deployment.html) and [http api](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-stages.html), perhaps the concept of a default stage doesn't exist on the former?

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/stage.ts:
```diff
@@ -9,22 +12,120 @@ export interface IStage extends IResource {
+  /**
+   * The API to which this stage is associated
+   */
+  readonly api: IApi;
```
So for this case in particular, a lot of common methods like metric* (or grant* in future perhaps) on the `StageBase` require an API to be present. I think it should be fine to have api a requirement to import the stage.

nija-at replied:
I still feel a customer shouldn't be forced to bring something that they may actually not need.

But this is fine for now. We can change it later too.

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/http/stage.ts:
```diff
@@ -38,77 +36,51 @@ export interface HttpStageProps extends HttpStageOptions {
 
   public readonly stageName: string;
-  private httpApi: IHttpApi;
+  public readonly api: IApi;
```
I agree on the ability to obtain `IHttpApi` instance from `IHttpStage`. I went ahead with the last option of adding an additional property to the individual stages and renamed the one in base class to `baseApi` making it `protected`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 67b3bb5ad2a4f4bc97953fc55b6193cc67a2a9a4
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4efdad4d-5ed3-444d-896c-60a47be0afe8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/base/api.ts:
```diff
@@ -0,0 +1,50 @@
+import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
```
(Missed this in the last round) 

Quite confusing to have base as another subfolder of common - `lib/common/base/api.ts`. Maybe move all the internal base classes into `lib/common/base.ts`?

ayush987goyal replied:
Yeah makes sense. Done.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 5cffd46b707ec2d5c0126e5f60009343690d8c33
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d27cef6c-3d56-47a8-8dc2-6d89e1e450de%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at reviewed (changes requested):
Just one comment. This is close to being shipped.

Can we capture the final set of breaking changes per my comment here - https://github.com/aws/aws-cdk/pull/13031#pullrequestreview-598646778
For class name changes, just capturing them as a single change is sufficient, but we should capture most if not all property changes.

I've found [jsii-diff](https://github.com/aws/jsii/tree/main/packages/jsii-diff) helps to capture. You should be able to get the output by running `npx jsii-diff npm:`

ayush987goyal commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/api-mapping.ts:
```diff
@@ -11,3 +17,107 @@ export interface IApiMapping extends IResource {
+  constructor(scope: Construct, id: string, props: ApiMappingProps) {
+    super(scope, id);
+
+    let stage = props.stage;
+    if (!stage) {
+      if (props.api instanceof HttpApi) {
+        if (props.api.defaultStage) {
+          stage = props.api.defaultStage;
+        } else {
+          throw new Error('stage is required if default stage is not available');
+        }
+      } else {
+        throw new Error('stage is required for WebSocket API');
+      }
+    }
```
So the `api` prop would still be required since currently `api` is not available on `IStage` since we moved it to subclass `IHttpStage` and `IWebSocketStage`. Now given that, we can keep this implementation because it provided convenience for HttpApi users (though agree this is not the cleanest of solutions but maybe in future it can be clear when we have defaultStage for WS apis as well). Thoughts?

nija-at replied:
So now we can go from `IHttpStage` -> `IHttpApi` but can't go from `IStage` -> `IApi`?  Whoops.
I think we should somehow get both but let's not try to solve this now in this PR.

Leave it as such.

ayush987goyal commented:
@nija-at Here are the generated changes by `jsii-diff` from which I updated the PR description. Let me know if something needs to be added removed:

```
Original assembly: @aws-cdk/aws-apigatewayv2@1.91.0
Updated assembly:  @aws-cdk/aws-apigatewayv2@0.0.0
API elements with incompatible changes:
warn - CLASS @aws-cdk/aws-apigatewayv2.HttpApiMapping: has been removed
warn - IFACE @aws-cdk/aws-apigatewayv2.CommonStageOptions: has been removed
warn - IFACE @aws-cdk/aws-apigatewayv2.DefaultDomainMappingOptions: has been removed
warn - IFACE @aws-cdk/aws-apigatewayv2.HttpApiMappingAttributes: has been removed
warn - IFACE @aws-cdk/aws-apigatewayv2.HttpApiMappingProps: has been removed
warn - PROP @aws-cdk/aws-apigatewayv2.HttpApi.defaultStage: type Optional<@aws-cdk/aws-apigatewayv2.IStage> (formerly Optional<@aws-cdk/aws-apigatewayv2.HttpStage>): @aws-cdk/aws-apigatewayv2.IStage does not extend @aws-cdk/aws-apigatewayv2.HttpStage
warn - METHOD @aws-cdk/aws-apigatewayv2.HttpStage.fromStageName: has been removed
warn - IFACE @aws-cdk/aws-apigatewayv2.DomainMappingOptions: not assignable to all base types anymore: could not find type @aws-cdk/aws-apigatewayv2.DefaultDomainMappingOptions
warn - PROP @aws-cdk/aws-apigatewayv2.DomainMappingOptions.stage: has been removed
warn - IFACE @aws-cdk/aws-apigatewayv2.HttpApiProps: property defaultDomainMapping, could not find type @aws-cdk/aws-apigatewayv2.DefaultDomainMappingOptions: input to @aws-cdk/aws-apigatewayv2.HttpApi.<initializer>
warn - IFACE @aws-cdk/aws-apigatewayv2.HttpStageOptions: not assignable to all base types anymore: could not find type @aws-cdk/aws-apigatewayv2.CommonStageOptions
warn - IFACE @aws-cdk/aws-apigatewayv2.HttpStageProps: not assignable to all base types anymore: could not find type @aws-cdk/aws-apigatewayv2.CommonStageOptions
warn - PROP @aws-cdk/aws-apigatewayv2.IHttpApi.defaultStage: has been removed
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f9494303ac4e0701732cbca49d4aae5d581b0769
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=83f400bc-b6a0-4911-8e73-74fb997b289a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented on packages/@aws-cdk/aws-apigatewayv2/lib/common/api-mapping.ts:
```diff
@@ -11,3 +17,109 @@ export interface IApiMapping extends IResource {
+   * required for WebSocket API
+   * defaults to default stage of an HTTP API
+   *
+   * @default - Default stage of the passed API for HTTP API, undefined for WebSocket API
```
```suggestion
   * @default - Default stage of the passed API for HTTP API, required for WebSocket API
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 1a1f0cf4455200c953c6072c677ef319a0af21e0
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=52f3d033-1738-4c6d-a248-b00e8abf617d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

ayush987goyal commented:
Hi @nija-at , is there anything left on this change? Can we merge it?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 75f7e5e2bcecb5e3e24dd4a7edb9d896a623b96c
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b279f8bc-a105-429d-a461-cdd0b545a14c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

ayush987goyal commented:
@nija-at Not sure what the above error is. Seems intermittent. I think we might have to retry.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: cf16610b271adc69b28ca1ede902e5c396f231f2
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b135dc6f-3860-46d1-a270-7c2dbd43127c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nija-at commented:
```
yarn install v1.22.5
[1/4] Resolving packages...
[2/4] Fetching packages...
error An unexpected error occurred: "https://registry.yarnpkg.com/debug/-/debug-4.3.1.tgz: Request failed \"500 Internal Server Error\"".
```

I'll restart the build

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 75f7e5e2bcecb5e3e24dd4a7edb9d896a623b96c
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f90ebdd2-6ef3-406b-a24e-b57dfe892864%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
